File: src/pages/HomePage.ts
import { Page, Locator } from '@playwright/test';
import { findSelfHealingElement } from '../core/PlaywrightExtensions';

export class HomePage {
  constructor(private page: Page) {}

  private async getLoginLink(): Promise<Locator> {
    return await findSelfHealingElement(this.page, 'text=Login');
  }

  private async getEmployeeDetailsLink(): Promise<Locator> {
    return await findSelfHealingElement(this.page, 'text=Employee Details');
  }

  private async getManageUsersLink(): Promise<Locator> {
    return await findSelfHealingElement(this.page, 'text=Manage Users');
  }

  private async getLogoffLink(): Promise<Locator> {
    return await findSelfHealingElement(this.page, 'text=Log off');
  }

  async clickLogin(): Promise<void> {
    const link = await this.getLoginLink();
    await link.click();
  }

  async clickEmployeeDetails(): Promise<void> {
    const link = await this.getEmployeeDetailsLink();
    await link.click();
  }

  async clickManageUsers(): Promise<void> {
    const link = await this.getManageUsersLink();
    await link.click();
  }

  async clickLogoff(): Promise<void> {
    const link = await this.getLogoffLink();
    await link.click();
  }
}

File: SeleniumSelfHealingTests/Pages/HomePage.cs
using OpenQA.Selenium;
using SeleniumSelfHealingTests.Utilities.Extensions;

namespace SeleniumSelfHealingTests.Pages;

public class HomePage
{
    private readonly IWebDriver _driver;
    private Task<IWebElement> LoginLink => _driver.AiFindElement(By.LinkText("Login"));
    private Task<IWebElement> EmployeeDetailsLink => _driver.AiFindElement(By.LinkText("Employee Details"));
    private Task<IWebElement> ManageUsersLink => _driver.AiFindElement(By.LinkText("Manage Users"));
    private Task<IWebElement> LogoffLink => _driver.AiFindElement(By.LinkText("Log off"));
    
    public HomePage(IWebDriver driver)
    {
        _driver = driver;
    }

    // One-step click operations using the element properties and new extension methods
    public async Task ClickLogin() => await LoginLink.Click();
    
    public async Task ClickEmployeeDetails() => await EmployeeDetailsLink.Click();
    
    public async Task ClickManageUsers() => await ManageUsersLink.Click();
    
    public async Task ClickLogoff() => await LogoffLink.Click();
}
